### DISCUSSION TITLE
Sensorperiod

### DISCUSSION THREAD

**Discussion by jeroenst** [arendst/Tasmota#16640] [Q&A] [answered]
I use several power monitoring devices (like Sonoff Pow R2 and Athom Plug) but the value is reported every second on a washer when using powerdelta due to the big changes in power usage.

I could also use teleperiod 5 for example but that would report every 5 seconds, also when there is no change.

My idea is to implement sensorperiod which sets the minimum time between 2 mqtt  sensor messages.

Is this a good idea or is there another way to prevent those many mqtt power messages?

**Comment by barbudor** [comment]
Powerdelta can already be set to trigger on either an absolute delta or a percentage delta
Doesn't it provide you already enough flexibility?
If you have too many messages, may be your threshold is too low

What is your exact objective with getting messages on changes but not too often?

**Reply by jeroenst** [reply]
Well the washer jumps from 6 watts to 60 to 130 and back so I would have to set a big power delta and only receive major changes.

The goal is to reduce network, server and client load due to many messages from all power monitoring devices. 

An update every 5 seconds when power has changed would be fine for me.

**Reply by barbudor** [reply]
> jumps from 6 watts to 60 to 130

Isn't it a major change worth of being notified ?

You should be able to do something with Rules but that would probably need a self-compiled with [SUPPORT_IF_STATEMENT](https://tasmota.github.io/docs/Rules/#conditional-rules)

IMHO It's not worth
I have `PowerDelta 105` on my washer's plug with `Teleperiod 3600`.
A lot of points when the washer is running but I don't care


![image](https://user-images.githubusercontent.com/2996042/192093583-83f60cd2-959f-49a4-b574-e351876f017a.png)
![image](https://user-images.githubusercontent.com/2996042/192093602-6263b14b-529f-4df1-b740-dcdcc1f584ec.png)

**Reply by sfromis** [reply]
If you do not want such details, do not set options asking to get details. And no, you could not set TelePeriod 5 seconds, as the minimum is already 10 seconds.

In case of wanting to micro-manage details of how energy monitoring sensors are reported, I'd use an energy monitoring device with ESP32 (both Shelly and Sonoff have such), use the [energy](https://tasmota.github.io/docs/Berry/#energy-module) module in Berry to follow values, and then instead of using `PowerDelta` issue commands to trigger reporting when wanted. This would give full flexibility to decide what I wanted.

Anyway, I'd be more likely to filter it backend, as the MQTT traffic is not a concern, and a little bit of processing before pushing values into the database could do whatever.

**Reply by jeroenst** [reply]
Ok how about remote locations with ssl? 

I have pretty much of them. Why sending a huge amount of unnecessary updates which are not useful only using bandwidth and filing database space without filtering. 

Imo every an update every second is often not preferred.

If agreed I can make a PR for this.

**Reply by barbudor** [reply]
Did you tried to do it per rule as I suggested ?

**Reply by jeroenst** [reply]
Not yet, for now I accept the extra load, hoping a rate limit for sending sensor updates is aproved here...

**Reply by jeroenst** [reply]
Sometimes the updates are send every 500ms to the MQTT server, isn't this to often? Especially when there are 50 tasmota devices which power sensor are connected over the internet?

When setting powerdelta to high no updates are send regularly when the load changes a little bit.

When I could combine a high powerdelta of 50 with a sensorperiod of 10 the server would be updated immediately on big power changes but would report minor power changes only once every 10 seconds.

It would be great if this command would be part of tasmota, otherwise using a rule that would send an update every 10 seconds would do also the job, but I have to investigate how to make such rule.

**Reply by sfromis** [reply]
Again: If you do not want such details, do not set options asking to get such details. Or use a device with ESP32 for greatly enhanced local processing options.

**Reply by jeroenst** [reply]
I definitely want those details but not in a 500ms minimum period.

**Reply by sfromis** [reply]
So you do not want the details you asked for.

**Reply by jeroenst** [reply]
I want an update every 5 seconds only if power has changed (also minimal changes of 1 watt), this causes less traffic (and without filtering less datastorage and less dataprocessing).

**Reply by jeroenst** [reply]
With an appliance like a bulb or television there is no problem using only powerdelta because the load is constant. The washing machine changes in power very often, it switches between turning clockwise and counterclockwise constantly, this causes a lot of messages without a sensorperiod. I now have set the powerdelta to 25. That works ok, but still I would rather see a rate limiting function on the messages send to the mqtt broker for sensor updates in this case.

**Reply by barbudor** [reply]
Because not everything needs a PR and you didn't seems to make the effort to try Rules, here is the solution
```
Rule3
  On system#boot do backlog var5 0; var6 0; var7 0 endon
  On Energy#Power>%var5% do event changed=%value% endon 
  On Energy#Power<%var6% do event changed=%value% endon 
  On Event#Changed do if (%var7%==0) var7 1; RuleTimer7 5; Teleperiod; var5 %value%; var6 %value%; add5 5; sub6 5 endif endon 
  On Rules#Timer=7 do var7 0 endon
backlog var5 0; var6 0; var7 0
Rule3 1
```
Same as `PowerDelta 105` with a publish limitation to 5 seconds

**Reply by barbudor** [reply]
Works fine on my Gosund plug

**Reply by jeroenst** [reply]
Thank you very much! I'm going to test it this week when the machine is running. 

I actually had started creating a rule but it isn't so easy for me.

**Reply by barbudor** [reply]
Don't forget, to use IF in rules you need to self compile as per the doc

**Comment by jeroenst** [comment]
Now I have a shelly 3em which I would like to report every 5 seconds. It now reports every second because of a low
powerdelta. I want an update every 5 seconds, so would it be an idea to implement sensorperiod which can be set for publishing sensor data. Teleperiod can't be shorter than 10 seconds and it is not needed to publish other data than only the sensor data so often.

**Reply by sfromis** [reply]
Well, sounds like you have customized PowerDelta to being too sensitive then.

More generally, if you have esoteric expectations, you have rules, etc. Or just stop micromanaging things not harming.

**Comment by jeroenst** [comment]
I tried rules but that also gives me 2 mqtt messages, one for the result of status 10 and 1 for setting the ruletimer again.

Powerdelta just doesn't match my needs and teleperiod 10 gives 2 messages and is not often enough for me.

Ideally I would get powerchanges of 50 or 100 watt directly and a regular update every 5 seconds.

I could make that with a rule but a command like sensorperiod would generate only 1 messages instead of 2 and I only have to parse the sensor topic instead of 2 topic.

**Reply by jeroenst** [reply]
I now use this rule :

Rule1 on system#boot do RuleTimer1 5 endon
on rules#timer=1 do backlog status 10; ruletimer1 5 endon

This works but generates 2 messages every 5 seconds. Is there a way to not send the result of setting the ruletimer?

**Reply by jeroenst** [reply]
```
21:35:31.339 SRC: Rule
21:35:31.342 CMD: Grp 0, Cmd 'BACKLOG', Idx 1, Len 23, Pld -99, Data 'status 10; ruletimer1 5'
21:35:31.371 SRC: Backlog
21:35:31.374 CMD: Grp 0, Cmd 'STATUS', Idx 1, Len 2, Pld 10, Data '10'
21:35:31.416 MQT: home/TASMOTA-SHELLY3EM-01/stat/STATUS10 = {"StatusSNS":{"Time":"2024-12-11T21:35:31","ENERGY":{"TotalStartTime":"2024-11-30T13:18:05","Total":[88.104,110.754,0.092],"Yesterday":[3.698,3.703,0.005],"Today":[8.544,8.776,0.010],"TodaySumImport":0.260,"TodaySumExport":0.000,"ExportActive":[17.743,0.000,0.028],"Power":[332,334,0],"ApparentPower":[524,521,3],"ReactivePower":[406,400,3],"Factor":[0.63,0.64,0.11],"Frequency":[50,50,50],"Voltage":[246,246,246],"Current":[2.143,2.129,0.014],"CurrentNeutral":0.011}}} (retained)
21:35:31.581 SRC: Backlog
21:35:31.585 CMD: Grp 0, Cmd 'RULETIMER', Idx 1, Len 1, Pld 5, Data '5'
21:35:31.626 MQT: home/TASMOTA-SHELLY3EM-01/stat/RESULT = {"T1":5,"T2":0,"T3":0,"T4":0,"T5":0,"T6":0,"T7":0,"T8":0}
```

**Reply by sfromis** [reply]
You can prefix command names with `_` to avoid getting the result output, on both console and MQTT.

**Reply by jeroenst** [reply]
Thank you very much!

**Comment by jeroenst** [comment, marked as answer]
Sensorperiod would be nice to have but this is good enough for me. Thank you for support. 

```
Rule1 on system#boot do RuleTimer1 5 endon
on rules#timer=1 do backlog status 10; _ruletimer1 5 endon
Rule1 1
```
